Support public key derivation in CLI

Anviking opened the pull request:
## Issue Number

ADP-190, #1594


## Overview

<!-- Detail in a few bullet points the work accomplished in this PR -->

- [x] Add new `XPrvOrXPub` type
- [x] Add hex coders for `XPrvOrXPub`
- [x] Test `derive . toPublic == toPublic . derive (for soft indices)`
- [x] Implement `inspect` and `toPublic` using the `CliKeyScheme` instead of separate functions
- [x] `derive` and `inspect` now accepts public keys as well
- [x] Extend roundtrip tests to test `inspect` and `toPublic`
- [ ] <s>Remove some boilerplate</s>
- [ ] <s>Consider parameterising by `prv pub any` instead of just `key` for slightly more helpful error messages.</s>
- [ ] <s>Maybe `Data.Codec` could be used?</s>

## Comments

<!-- Additional comments or screenshots to attach if any -->

<!-- 
Don't forget to:

 ✓ Self-review your changes to make sure nothing unexpected slipped through
 ✓ Assign yourself to the PR
 ✓ Assign one or several reviewer(s)
 ✓ Once created, link this PR to its corresponding ticket
 ✓ Assign the PR to a corresponding milestone
 ✓ Acknowledge any changes required to the Wiki
-->

Anviking commented on lib/cli/src/Cardano/CLI.hs:
```diff
@@ -400,27 +405,38 @@ hoistKeyScheme eta s = CliKeyScheme
-        left showErr $ xPrvFromStrippedPubXPrvCheckRoundtrip bytes
+        case BS.length bytes of
+            96 -> fmap AXPrv $ left showErr $ xPrvFromStrippedPubXPrvCheckRoundtrip bytes
+            64 -> fmap AXPub . CC.xpub $ bytes
```
Should use named constant

KtorZ replied:
This function really looks like it should be two functions :no_mouth: ...

KtorZ replied:
See for instance:

- https://input-output-hk.github.io/cardano-addresses/haddock/Cardano-Address-Derivation.html#v:xpubFromBytes
- https://input-output-hk.github.io/cardano-addresses/haddock/Cardano-Address-Derivation.html#v:xprvFromBytes

Anviking replied:
I intentionally wanted a single coder for `XPrvOrXPub` instead of ones for `XPub` and `XPrv` individually.

This way the "is the key private or public?" question `key child` has to answer can be handled by the coder and tested via roundtrips.

Anviking replied:
Or are you referring to the fact that it is written as a tuple `(encode, decode)`?

I wanted something more opaque than
```
mapKey (encodeHex, decodeHex) scheme
```

since it is important that the transformation is bidirectional

That is why it was called:
```
mapKey xPrvToTextTransform scheme
```

need to rename it now, maybe
```
mapKey hexKeyEncoding scheme
```
or something 🤷‍♂️

KtorZ replied:
I don't know, it feels wrong to have a single function that decodes an xpub or xprv, I'd expect two here since, whether you want one or the other is clear from the context where these functions are called. Although I get your point, if we want one common interface where we can just call `cardano-wallet key child` irrespective of whether the key is private or public then, it's not clear from the surrounding context anymore which decoder you should call. 

Still, I'd find it nicer to go for an alternative, trying one, and then the other.

Anviking replied:
> Still, I'd find it nicer to go for an alternative, trying one, and then the other.

`<|>` gives you less control over error messages, so I'm trying to avoid it.

>  I'd expect two here since

Well, not needed. Unless we differentiated between `xprv xpub any` and had separate encodings for all three. This way `key public` could avoid suggesting public keys are accepted 
```
$ cardano-wallet-byron key public 0000
Expected key to be 96 bytes in the case of a private key and, 64 bytes for public keys. This key is 2 bytes.
```
but then not:
```
$ cardano-wallet-byron key public 00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
Input is already a public key.
```

But since it is only one command, and fairly self-explanatory I thought it would be overkill.

Anviking commented on lib/cli/src/Cardano/CLI.hs:
```diff
@@ -587,18 +611,58 @@ newCliKeyScheme = \case
+            ]
+    insp (AXPub key) = do
+        let bytes = CC.unXPub key
+        let (xpub, cc) = BS.splitAt 32 bytes
```
Need to check that this is right.

Anviking replied:
Yes https://github.com/input-output-hk/cardano-crypto/blob/2547ad1e80aeabca2899951601079408becbc92c/src/Cardano/Crypto/Wallet.hs#L144-L150

Anviking committed:
Make CLIKeyScheme also accept public keys

- Add XPrvOrXPub
- Move the `key public` and `key inspect` implementations to
  CliKeyScheme.
- Make `key child` work on public keys for soft indices

Anviking committed:
Test `derive . toPublic == toPublic . derive (for soft indices)`

Anviking committed:
Extend `propCliKeySchemeEquality` to also check inspect and public

Anviking committed:
Add goldens for `key public` and `key inspect`

KtorZ commented on lib/cli/src/Cardano/CLI.hs:
```diff
@@ -400,27 +405,38 @@ hoistKeyScheme eta s = CliKeyScheme
         bytes <- fromHex $ B8.pack $ T.unpack . T.strip $ txt
-        left showErr $ xPrvFromStrippedPubXPrvCheckRoundtrip bytes
+        case BS.length bytes of
+            96 -> fmap AXPrv $ left showErr $ xPrvFromStrippedPubXPrvCheckRoundtrip bytes
```
Gotta love these names.

KtorZ commented on lib/cli/src/Cardano/CLI.hs:
```diff
@@ -587,18 +611,58 @@ newCliKeyScheme = \case
+        err = mconcat
+            [ T.unpack (toText i)
+            , " is a hardened index. Public key derivation is only possible for"
+            , " soft indices. \nIf the index is correct, please use the"
```
Small consistency tweak. This message uses both "indices" and "index" interchangeably, perhaps it's good to stick to one consistent wording otherwise people may start to wonder what's the difference between an indice and an index.

Anviking replied:
I believe both `indices` and `indexes` are valid plural forms of `index`

- https://www.nasdaq.com/articles/indexes-or-indices-whats-the-deal-2016-05-12
- https://developer.apple.com/documentation/swift/collection/1641719-indices

Actually we are already using both even without this PR (though indices much less):

```
git grep "indices"
.travis.yml:  - rm -rf $HOME/.stack/indices/Hackage/00-index.tar $HOME/.stack/indices/Hackage/00-index.tar.gz
lib/cli/src/Cardano/CLI.hs:            , " soft indices. \nIf the index is correct, please use the"
lib/cli/test/unit/Cardano/CLISpec.hs:                    \derivation is only possible for soft indices. \nIf the\
lib/cli/test/unit/Cardano/CLISpec.hs:        it "derive . toPublic == toPublic . derive (for soft indices)"
lib/cli/test/unit/Cardano/CLISpec.hs:-- | For soft indices, public key derivation should be "equivalent" to private
lib/core/src/Cardano/Wallet/DB/Sqlite.hs:    -- Note: there are sorted indices on these columns.
lib/core/src/Cardano/Wallet/DB/Sqlite/TH.hs:-- Mapping of pool addresses to indices, and the slot
lib/core/src/Cardano/Wallet/Primitive/AddressDerivation/Byron.hs:    -- ^ The address derivation indices for the level of this key.
lib/core/src/Cardano/Wallet/Primitive/AddressDerivation/Byron.hs:-- | The hierarchical derivation indices for a given level/depth.
lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Random.hs:    -- that addresses will be _discovered_ from any and all account indices,

git grep "indexes"
lib/byron/test/data/cardano-node/genesis.yaml:  # Address indexes below are #501, #502, ..., #510
lib/byron/test/integration/Test/Integration/Byron/Scenario/CLI/Addresses.hs:        describe "CLI_ADDRESS_CREATE_07 - False indexes" $ do
lib/byron/test/unit/Cardano/Wallet/Byron/TransactionSpec.hs:    <*> scale (`mod` 3) arbitrary -- No need for a high indexes
lib/cli/src/Cardano/CLI.hs:                , "Please use \"H\" to denote hardened indexes. "
lib/cli/test/unit/Cardano/CLISpec.hs:                     \Please use \"H\" to denote hardened indexes. Did you \
lib/core-integration/src/Test/Integration/Scenario/API/Byron/Wallets.hs:    it "BYRON_RESTORE_08 - Icarus wallet with high indexes" $ \ctx -> do
lib/core-integration/src/Test/Integration/Scenario/API/Byron/Wallets.hs:        -- Special Icarus mnemonic where address indexes are all after the index
lib/core/src/Cardano/Byron/Codec/Cbor.hs:                , "expected two indexes but got: "
lib/core/src/Cardano/Wallet/DB/Sqlite.hs:                -- wallet backend was generating addresses indexes across the
lib/core/src/Cardano/Wallet/DB/Sqlite.hs:                -- We later changed that to instead use "hardened indexes". Yet,
lib/core/src/Cardano/Wallet/DB/Sqlite/TH.hs:-- Sequential address discovery scheme -- pending change indexes
lib/core/src/Cardano/Wallet/Primitive/AddressDerivation.hs:-- | Type of derivation that should be used with the given indexes.
lib/core/src/Cardano/Wallet/Primitive/AddressDerivation/Shelley.hs:-- indexes and soft (non-hardened) indexes for various level in a non-ambiguous
lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Sequential.hs:-- in order to discover addresses with indexes separated by possible huge gaps.
lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Sequential.hs:-- expected to be ordered in sequence (first indexes, first in the list).
lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Sequential.hs:-- | An ordered set of pending indexes. This keep track of indexes used
lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Sequential.hs:-- | An empty pending set of change indexes.
lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Sequential.hs:-- | Update the set of pending indexes by discarding every indexes _below_ the
lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Sequential.hs:-- to know from where we can generate new indexes.
lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Sequential.hs:-- | Construct a 'PendingIxs' from a list, ensuring that it is a set of indexes
lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Sequential.hs:-- | Get the next change index; If every available indexes have already been
lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Sequential.hs:-- taken, we'll rotate the pending set and re-use already provided indexes.
lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Sequential.hs:        <> indentF 4 ("Change indexes: " <> indentF 4 chgsF)
lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Sequential.hs:    -- | We pick indexes in sequence from the first known available index (i.e.
lib/core/src/Cardano/Wallet/Primitive/Types.hs:        -- their indexes. It matters also for serialization.
lib/core/test/unit/Cardano/Wallet/Api/TypesSpec.hs:        -- NOTE: No need for a crazy high indexes
lib/core/test/unit/Cardano/Wallet/DB/Arbitrary.hs:        <*> scale (`mod` 3) arbitrary -- No need for a high indexes
lib/core/test/unit/Cardano/Wallet/DB/Arbitrary.hs:-- indexes. For example:
lib/core/test/unit/Cardano/Wallet/DB/Arbitrary.hs:-- - put checkpoint with pending indexes
lib/core/test/unit/Cardano/Wallet/DB/Arbitrary.hs:-- - put checkpoint with empty pending indexes
lib/core/test/unit/Cardano/Wallet/DB/Arbitrary.hs:-- currently, it'll also contains the pending indexes from the previously
lib/core/test/unit/Cardano/Wallet/Primitive/AddressDerivation/ByronSpec.hs:        it "Key derivation from seed works for various indexes" $
lib/core/test/unit/Cardano/Wallet/Primitive/AddressDerivation/ByronSpec.hs:        it "Key derivation from master key works for various indexes" $
lib/core/test/unit/Cardano/Wallet/Primitive/AddressDerivation/IcarusSpec.hs:        it "deriveAccountPrivateKey works for various indexes" $
lib/core/test/unit/Cardano/Wallet/Primitive/AddressDerivation/ShelleySpec.hs:        it "deriveAccountPrivateKey works for various indexes" $
lib/core/test/unit/Cardano/Wallet/Primitive/AddressDiscovery/RandomSpec.hs:    it "findUnusedPath: indexes are always in the 'hardened' realm" $
lib/core/test/unit/Cardano/Wallet/Primitive/CoinSelectionSpec.hs:        <*> scale (`mod` 3) arbitrary -- No need for a high indexes
lib/core/test/unit/Cardano/Wallet/Primitive/FeeSpec.hs:        <*> scale (`mod` 3) arbitrary -- No need for a high indexes
lib/core/test/unit/Cardano/Wallet/Primitive/TypesSpec.hs:        <*> scale (`mod` 3) arbitrary -- No need for a crazy high indexes
lib/jormungandr/test/data/jormungandr/genesis.yaml:  # Address indexes below are #501, #502, ..., #510
lib/jormungandr/test/integration/Main.hs:-- outputs and the values of the input indexes.
```

ouroboros-network has more occurrences of indices than indexes though

😅 🔫

KtorZ commented on lib/cli/test/unit/Cardano/CLISpec.hs:
```diff
@@ -809,17 +855,58 @@ prop_roundtripCliKeySchemeKeyViaHex :: ByronWalletStyle -> Property
+--                v                       v
+--           child xprv +-----------> child xpub
+--
+--                        toPublic
```
Noice :heart_eyes:

KtorZ commented on lib/cli/test/unit/Cardano/CLISpec.hs:
```diff
@@ -809,17 +855,58 @@ prop_roundtripCliKeySchemeKeyViaHex :: ByronWalletStyle -> Property
+    -> Property
+prop_publicKeyDerivation style i key = do
+    if isSoftIndex i
+    then (public k >>= derive) === (derive k >>= toPublic s)
```
Noice x2

KtorZ reviewed (approved):
:+1:
